File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    min_time: int = 1
    max_time: int = max(times) * n
    
    while min_time <= max_time:
        mid: int = (min_time + max_time) // 2
        count: int = 0
        for time in times:
            count += mid // time
        if count >= n:
            max_time = mid - 1
        else:
            min_time = mid + 1
    
    return min_time

File: 프로그래머스/3/43238. 입국심사/입국심사.java
class Solution {
    public long solution(int n, int[] times) {
        long minTime = Long.MAX_VALUE;
        for (long time : times) {
            if (time < minTime) minTime = time;
        }
        
        long left = 0L;
        long right = (long) minTime * n;
        long answer = 0;
        while (left <= right) {
            long mid = (long) (left + right) / 2;
            
            if (can(mid, times, n)) {
                answer = mid;
                right = mid - 1;
            } else left = mid + 1;
            
        }
        return answer;
    }
    
    private boolean can(long mid, int[] times, long n) {
        long done = 0L;
        for (int time : times) {
            done += mid / time;
            if (done >= n) return true;
        }
        return false;
    }
}